### PR TITLE
Add authentication and registration logging

### PR DESCRIPTION
**Pull request by epinxteren**
Add login for authentication and registration flows.

I have added comments to explain some changes.

- Add test env, with overridden logger.
- Fixed composer.json phpmd, dit not work.

### CODE REVIEW

File: ansible/templates/nginx-vhost.conf.j2 [hunk 1]
```diff
@@ -27,7 +27,7 @@ server {
         try_files $uri /app.php$is_args$args;
     }
 
-    location ~ ^/app(_dev)?\.php(/|$) {
+    location ~ ^/app(_(dev|test))?\.php(/|$) {
```
Comment (epinxteren) [inline ansible/templates/nginx-vhost.conf.j2]: Add support for test env

File: src/AppBundle/Controller/AuthenticationController.php [hunk 2]
```diff
@@ -29,74 +30,127 @@ class AuthenticationController extends Controller
...
         }
 
+
         // Handle one time password
         if ($request->get('otp') !== null) {
+            $this->logger->info('authenticationAction: handling otp', $logContext);
+
             // Implement otp
-            return new Response('One time passord: '. $request->get('otp'), Response::HTTP_I_AM_A_TEAPOT);
+            return new Response('One time passord: '.$request->get('otp'), Response::HTTP_I_AM_A_TEAPOT);
```
Comment (epinxteren) [inline src/AppBundle/Controller/AuthenticationController.php]: This is next feature i'm working on, adding support for one time passwords

File: src/AppBundle/Controller/AuthenticationController.php [hunk 3]
```diff
@@ -29,74 +30,127 @@ class AuthenticationController extends Controller
...
+        $this->logger = $logger;
     }
 
     /**
      * @Route("/authentication", name="app_identity_authentication")
+     * @throws \InvalidArgumentException
      */
     public function authenticationAction(Request $request)
     {
+        $this->logger->info('authenticationAction: Verifying if there is a pending authentication request from SP');
```
[the remark this answers is not in the document]
Comment (epinxteren) [reply src/AppBundle/Controller/AuthenticationController.php]: Removed  action/method name. Did not add variables, does not have them in this scope.

File: src/AppBundle/Controller/AuthenticationController.php [hunk 4]
```diff
@@ -29,74 +30,127 @@ class AuthenticationController extends Controller
...
             return $this->authenticationService->replyToServiceProvider();
         }
 
         // Start authentication process.
         try {
+            $this->logger->info('authenticationAction: start authentication', $logContext);
             $this->tiqrService->startAuthentication($nameId);
         } catch (\Exception $e) {
+            $this->logger->error(sprintf(
+                'authenticationAction: Failed to start authentication %s',
```
[the remark this answers is not in the document]
Comment (epinxteren) [reply src/AppBundle/Controller/AuthenticationController.php]: 👍

File: src/AppBundle/Controller/AuthenticationController.php [hunk 5]
```diff
@@ -29,74 +30,127 @@ class AuthenticationController extends Controller
...
         // Handle one time password
         if ($request->get('otp') !== null) {
+            $this->logger->info('authenticationAction: handling otp', $logContext);
+
             // Implement otp
-            return new Response('One time passord: '. $request->get('otp'), Response::HTTP_I_AM_A_TEAPOT);
+            return new Response('One time passord: '.$request->get('otp'), Response::HTTP_I_AM_A_TEAPOT);
         }
 
+        $this->logger->info('authenticationAction: Verifying if authentication is finalized', $logContext);
```
[the remark this answers is not in the document]
Comment (epinxteren) [reply src/AppBundle/Controller/AuthenticationController.php]: 👍

File: src/AppBundle/Controller/AuthenticationController.php [hunk 6]
```diff
@@ -29,74 +30,127 @@ class AuthenticationController extends Controller
...
         try {
+            $this->logger->info('authenticationAction: start authentication', $logContext);
             $this->tiqrService->startAuthentication($nameId);
         } catch (\Exception $e) {
+            $this->logger->error(sprintf(
+                'authenticationAction: Failed to start authentication %s',
+                $e->getMessage()
+            ), $logContext);
+            $this->logger->info(
+                'authenticationAction: Return authentication failed response',
```
[the remark this answers is not in the document]
Comment (epinxteren) [reply src/AppBundle/Controller/AuthenticationController.php]: 👍 Fixed in multiple places

File: src/AppBundle/Controller/AuthenticationController.php [hunk 7]
```diff
@@ -29,74 +30,127 @@ class AuthenticationController extends Controller
...
+                $logContext
+            );
             $this->authenticationService->reject($e->getMessage());
 
             return $this->authenticationService->replyToServiceProvider();
         }
 
+        $this->logger->info(
+            'authenticationAction: Return authentication page with qr code',
```
[the remark this answers is not in the document]
Comment (epinxteren) [reply src/AppBundle/Controller/AuthenticationController.php]: Fixed in many places

File: src/AppBundle/Controller/TqirAppApiController.php [hunk 8]
```diff
@@ -146,79 +180,142 @@ private function registerAction(Request $request, $notificationType, $notificati
...
         } catch (UserNotExistsException $e) {
+            $this->logger->error('loginAction: User does not exists', $logContext);
+
             return new Response('INVALID_USER', Response::HTTP_BAD_REQUEST);
         }
 
-        // Check if user is blocked. (TODO: maybe do this in AuthenticationController?)
         if ($this->configuration->temporaryBlockEnabled() &&
             $user->isBlockTemporary($now, $this->configuration->getTemporaryBlockDuration())) {
+            $this->logger->info('loginAction: User is temporary blocked', $logContext);
```
[the remark this answers is not in the document]
Comment (epinxteren) [reply src/AppBundle/Controller/TqirAppApiController.php]: I will do this in the next MR, this logic is refactored

File: src/AppBundle/Controller/TqirAppApiController.php [hunk 9]
```diff
@@ -146,79 +180,142 @@ private function registerAction(Request $request, $notificationType, $notificati
...
+        array $logContext,
+        AuthenticationResponse $result,
+        TiqrUserInterface $user
+    ) {
+        $now = new DateTimeImmutable();
+        $this->logger->info('loginAction: User login attempt is rejected', $logContext);
+
+        // If there is no limit how many times the user can try to login.
+        if (!$this->configuration->hasMaxLoginAttempts()) {
+            $this->logger->info('loginAction: Return forbidden and ignore attempt', $logContext);
```
[the remark this answers is not in the document]
Comment (epinxteren) [reply src/AppBundle/Controller/TqirAppApiController.php]: 👍

File: app/config/config_test.yml [hunk 10]
```diff
@@ -11,3 +11,19 @@ framework:
...
+        service_providers:
+            - entity_id: "%saml_remote_sp_entity_id%"
+              certificate_file: "%saml_remote_sp_certificate%"
+              assertion_consumer_service_url: "%saml_remote_sp_acs%"
+            - entity_id: https://tiqr.example.com/app_test.php/saml/metadata
+              certificate_file: "%saml_idp_publickey%"
+              assertion_consumer_service_url: https://tiqr.example.com/app_test.php/demo/sp/acs
+
+services:
+    AppBundle\Features\Context\FileLogger:
```
[the remark this answers is not in the document]
Comment (epinxteren) [reply app/config/config_test.yml]: Yes, that will log to file, so we can read in the Behat tests.

File: src/AppBundle/Controller/TqirAppApiController.php [hunk 11]
```diff
@@ -146,79 +180,142 @@ private function registerAction(Request $request, $notificationType, $notificati
...
+            $user->block();
+            $this->logger->info('loginAction: User is blocked indefinitely', $logContext);
+
+            return new Response('ACCOUNT_BLOCKED', Response::HTTP_FORBIDDEN);
+        }
+
+        // Just block the user temporary if we don't got a limit.
+        if (!$this->configuration->hasMaxTemporaryLoginAttempts()) {
+            $user->blockTemporary($now);
+            $this->logger->info('loginAction: Increase temporary block attempt', $logContext);
```
[the remark this answers is not in the document]
Comment (epinxteren) [reply src/AppBundle/Controller/TqirAppApiController.php]: 👍

File: src/AppBundle/Controller/TqirAppApiController.php [hunk 12]
```diff
@@ -34,21 +38,32 @@
...
+ *
+ * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ * @SuppressWarnings(PHPMD.NPathComplexity)
+ *
+ * Thought about converting the logic in loginAction into a state machine
+ * After some consideration with a college this would make sense if it get's even bigger for now it's
+ * still kind of readable.
  */
 class TqirAppApiController extends Controller
```
Comment (joostd) [inline src/AppBundle/Controller/TqirAppApiController.php]: TiqrAppApiController

Comment (epinxteren) [reply src/AppBundle/Controller/TqirAppApiController.php]: 👍

File: src/AppBundle/Controller/AuthenticationController.php [hunk 13]
```diff
@@ -29,74 +30,127 @@ class AuthenticationController extends Controller
...
+        $this->logger->info('Verifying if there is a pending authentication request from SP');
+
         $nameId = $this->authenticationService->getNameId();
+        $logContext = ['nameId' => $nameId];
+
 
         // Do have a valid sample AuthnRequest?.
         if (!$this->authenticationService->authenticationRequired()) {
+            $this->logger->error(
+                'there is no pending authentication request from SP',
```
Comment (jorissteyn) [inline src/AppBundle/Controller/AuthenticationController.php]: Lets start all messages upper case.

File: src/AppBundle/Controller/AuthenticationController.php [hunk 14]
```diff
@@ -29,74 +30,127 @@ class AuthenticationController extends Controller
...
         }
 
+
         // Handle one time password
         if ($request->get('otp') !== null) {
+            $this->logger->info('handling otp', $logContext);
+
             // Implement otp
-            return new Response('One time passord: '. $request->get('otp'), Response::HTTP_I_AM_A_TEAPOT);
+            return new Response('One time passord: '.$request->get('otp'), Response::HTTP_I_AM_A_TEAPOT);
```
Comment (jorissteyn) [inline src/AppBundle/Controller/AuthenticationController.php]: Typo: passWord

File: src/AppBundle/Controller/AuthenticationController.php [hunk 15]
```diff
@@ -29,74 +30,127 @@ class AuthenticationController extends Controller
...
+            $this->logger->error('there is no pending authentication request from SP');
+
             return new Response('No authentication required', Response::HTTP_BAD_REQUEST);
         }
 
-        return new JsonResponse($this->tiqrService->isAuthenticated());
+        $isAuthenticated = $this->tiqrService->isAuthenticated();
+
+        if ($isAuthenticated) {
+            $this->logger->info('Send json response is authenticated');
```
Comment (jorissteyn) [inline src/AppBundle/Controller/AuthenticationController.php]: You mean sent, nod send.

File: src/AppBundle/Controller/AuthenticationController.php [hunk 16]
```diff
@@ -29,74 +30,127 @@ class AuthenticationController extends Controller
...
+            return new JsonResponse(true);
+        }
+        $this->logger->info('Send json response is not authenticated');
```
Comment (jorissteyn) [inline src/AppBundle/Controller/AuthenticationController.php]: Idem, send -> sent.

File: src/AppBundle/Controller/AuthenticationController.php (hunk 15 above)
Comment (jorissteyn) [reply src/AppBundle/Controller/AuthenticationController.php]: A message can be authenticated, but that is not what you mean. Proposal: Authentication successful, sending JSON response.

File: src/AppBundle/Controller/AuthenticationController.php (hunk 16 above)
Comment (jorissteyn) [reply src/AppBundle/Controller/AuthenticationController.php]: Proposal: Authentication failed, sending JSON response.

File: src/AppBundle/Controller/TiqrAppApiController.php [hunk 17]
```diff
@@ -65,15 +80,24 @@ public function metadataAction(Request $request)
     {
         $key = $request->get('key');
         if (empty($key)) {
+            $this->logger->info('without enrollment');
```
Comment (jorissteyn) [inline src/AppBundle/Controller/TiqrAppApiController.php]: Maybe it's better to make this one and "with key" a little more verbose. A log with a message "with key" doesn't say much.

File: src/AppBundle/Controller/TiqrAppApiController.php [hunk 18]
```diff
@@ -146,79 +180,142 @@ private function registerAction(Request $request, $notificationType, $notificati
...
         } catch (UserNotExistsException $e) {
+            $this->logger->error('User does not exists', $logContext);
+
             return new Response('INVALID_USER', Response::HTTP_BAD_REQUEST);
         }
 
-        // Check if user is blocked. (TODO: maybe do this in AuthenticationController?)
         if ($this->configuration->temporaryBlockEnabled() &&
             $user->isBlockTemporary($now, $this->configuration->getTemporaryBlockDuration())) {
+            $this->logger->info('User is temporary blocked', $logContext);
```
Comment (jorissteyn) [inline src/AppBundle/Controller/TiqrAppApiController.php]: temporary -> temporarily

File: src/AppBundle/Controller/TiqrAppApiController.php [hunk 19]
```diff
@@ -146,79 +180,142 @@ private function registerAction(Request $request, $notificationType, $notificati
...
-                $user->block();
-                return new Response('ACCOUNT_BLOCKED', Response::HTTP_FORBIDDEN);
-            }
+            return $this->handleAuthenticationRejectResponse($logContext, $result, $user);
+        }
 
-            $user->blockTemporary($now);
-            $attemptsLeft = $this->configuration->getMaxTemporaryLoginAttempts() - $user->getTemporaryLoginAttempts();
+        $this->logger->error(
+            sprintf('Unexpected authentication error accorded "%s"', $result->getMessage()),
```
Comment (jorissteyn) [inline src/AppBundle/Controller/TiqrAppApiController.php]: accorded -> occurred

File: src/AppBundle/Controller/TiqrAppApiController.php [hunk 20]
```diff
@@ -146,79 +180,142 @@ private function registerAction(Request $request, $notificationType, $notificati
...
+            );
+
+            return new Response('ACCOUNT_BLOCKED', Response::HTTP_FORBIDDEN);
+        }
+
+        $user->blockTemporary($now);
+        $attemptsLeft = $this->configuration->getMaxTemporaryLoginAttempts() - $user->getTemporaryLoginAttempts();
+
+        $this->logger->info(
+            sprintf('Increase temporary login attempt. Attempts left %s', $attemptsLeft),
```
Comment (jorissteyn) [inline src/AppBundle/Controller/TiqrAppApiController.php]: Minor nitpick: %s -> %d

File: src/AppBundle/Controller/CancelController.php [hunk 21]
```diff
@@ -27,28 +28,39 @@ class CancelController extends Controller
...
     }
 
     /**
      * @Route("/cancel", name="app_cancel")
+     * @throws \InvalidArgumentException
      */
-    public function authenticationAction()
+    public function cancelAction()
     {
+        $this->logger->notice('User canceled the request');
```
Comment (jorissteyn) [inline src/AppBundle/Controller/CancelController.php]: Canceled -> Cancelled

File: src/AppBundle/Controller/TiqrAppApiController.php (hunk 17 above)
Comment (epinxteren) [reply src/AppBundle/Controller/TiqrAppApiController.php]: 👍

**Pull request merged**